### DISCUSSION TITLE
UI Testing Blazor Hybrid for Windows

### DISCUSSION THREAD

**Discussion by joshbartley** [dotnet/maui#22200] [General]
I'm working on a Blazor Hybrid app that will run Android, iOS, Mac, and Windows. I want to add UI testing to the process. Current recommendation is to use Appium. Appium works for Maui for Windows but not Blazor Hybrid for Windows because Appium relies on the WinDriver and you need to switch to the web context but that isn't supported. Next plan, act like it's a UWP app that has a WebView2 in it and connect over the debugger URL. After about a day of trying, this has failed too.

https://learn.microsoft.com/en-us/microsoft-edge/webview2/how-to/webdriver#step-4b-attaching-microsoft-edge-webdriver-to-a-running-webview2-app
Setting `Environment.SetEnvironmentVariable("WEBVIEW2_ADDITIONAL_BROWSER_ARGUMENTS", "--enable-features=msEdgeDevToolsWdpRemoteDebugging");` does not seem to do much before the Maui app starts on windows. 

```c#
		EdgeOptions eo = new EdgeOptions();
		eo.AddAdditionalEdgeOption("wdpAddress", "localhost:50080");
		eo.AddAdditionalEdgeOption("wdpProcessId", 42168); //setting this from the developer device manager at http://localhost:50080/msedge
		EdgeDriver e = new EdgeDriver(eo);
		Screenshot screenshot = e.GetScreenshot();
		screenshot.SaveAsFile("screenshot.png");
```		
This throws `OpenQA.Selenium.WebDriverException : The HTTP request to the remote WebDriver server for URL http://localhost:57914/session timed out after 60 seconds.` which doesn't make sense because the `webSocketDebuggerUrl` is set to `ws://localhost:50080` and may just be some bug on the EdgeDriver side? 

I'm stuck here and would appreciate the communities help or know if this is something on the roadmap. All the UI testing Appium examples are MAUI but not Hybrid apps and searching has not brought up much help in this.

**Comment by stormbreakermerika** [comment]
> I'm working on a Blazor Hybrid app that will run Android, iOS, Mac, and Windows. I want to add UI testing to the process. Current recommendation is to use Appium. Appium works for Maui for Windows but not Blazor Hybrid for Windows because Appium relies on the WinDriver and you need to switch to the web context but that isn't supported. Next plan, act like it's a UWP app that has a WebView2 in it and connect over the debugger URL. After about a day of trying, this has failed too.
> 
> https://learn.microsoft.com/en-us/microsoft-edge/webview2/how-to/webdriver#step-4b-attaching-microsoft-edge-webdriver-to-a-running-webview2-app Setting `Environment.SetEnvironmentVariable("WEBVIEW2_ADDITIONAL_BROWSER_ARGUMENTS", "--enable-features=msEdgeDevToolsWdpRemoteDebugging");` does not seem to do much before the Maui app starts on windows.
> 
> ```cs
> 		EdgeOptions eo = new EdgeOptions();
> 		eo.AddAdditionalEdgeOption("wdpAddress", "localhost:50080");
> 		eo.AddAdditionalEdgeOption("wdpProcessId", 42168); //setting this from the developer device manager at http://localhost:50080/msedge
> 		EdgeDriver e = new EdgeDriver(eo);
> 		Screenshot screenshot = e.GetScreenshot();
> 		screenshot.SaveAsFile("screenshot.png");
> ```
> 
> This throws `OpenQA.Selenium.WebDriverException : The HTTP request to the remote WebDriver server for URL http://localhost:57914/session timed out after 60 seconds.` which doesn't make sense because the `webSocketDebuggerUrl` is set to `ws://localhost:50080` and may just be some bug on the EdgeDriver side?
> 
> I'm stuck here and would appreciate the communities help or know if this is something on the roadmap. All the UI testing Appium examples are MAUI but not Hybrid apps and searching has not brought up much help in this.

Did you ever get this solved? I also have this similar question

**Reply by joshbartley** [reply]
@stormbreakermerika I have not.  I'm waiting on .net 9 to release to see if anything changed that will make this easier to support testing on the windows side either on purpose or accident.

**Reply by joshbartley** [reply]
I have not tried setting a global environment variable for remote debugging though. Previous attempt was only doing that in the test itself which maybe another process was getting it picked up.

https://github.com/microsoft/WinAppDriver/issues/1668#issuecomment-1014776117

**Reply by joshbartley** [reply]
Oh, I think it might work using the environment variable in the WinAppDriver above.. It's might require some wrapper around the Edge driver and Appiumn. I was just able to get a screenshot from the webview on windows this way.

```c#
EdgeOptions eo = new EdgeOptions();
eo.DebuggerAddress = "localhost:9222";
eo.UseWebView = true;
EdgeDriver e = new EdgeDriver(eo);
Screenshot screenshot = e.GetScreenshot();
screenshot.SaveAsFile("screenshot.png");
```

**Reply by stormbreakermerika** [reply]
ok just finally got around to setting Appium up.  I plan to use it under reqnroll so once I get to that driver portion I'll refer back to this post and also do some exploring of my own

**Reply by stormbreakermerika** [reply]
I don't disagree with other comments that this should be voted higher, I came to the conclusion that a hybrid app was not necessarily the way to go for me.  But you are correct in the workaround for needing to remote debug the app through a browser and run a browser driver that way

**Comment by akhanalcs** [comment]
This is important and should be voted higher.

**Comment by arnaringib** [comment]
Here is an example of how I got this working https://github.com/arnaringib/MAUI.Hybrid.Integration.Example
